src/aws-sdk/types.ts

```diff
@@ -0,0 +1,103 @@
+import type { Service as AWSService } from "aws-sdk";
+// eslint-disable-next-line import/no-extraneous-dependencies
+import type * as AWS from "aws-sdk";
+import { AnyFunction, OverloadUnion } from "../util";
+
+type AWSServiceClass = { new (): AWSService };
+
+/**
+ * First we have to extract the names of all Services in the v2 AWS namespace
+ *
+ * @returns "AccessAnalyzer" | "Account" | ... | "XRay"
+ */
+export type ServiceKeys = Exclude<
+  {
+    [K in keyof typeof AWS]: typeof AWS[K] extends AWSServiceClass ? K : never;
+  }[keyof typeof AWS],
+  "Service"
+>;
+
+export type SDK = {
+  [serviceName in ServiceKeys]: typeof AWS[serviceName] extends new (
+    ...args: any[]
+  ) => infer Client
+    ? {
+        [methodName in keyof Client]: SdkMethod<Client[methodName]>;
+      }
+    : never;
+};
+
+export interface SdkCallInput<Params extends object> {
+  /**
+   * The parameters to the api call
+   */
+  params: Params;
+
+  /**
+   * The resources for the IAM statement that will be added to the state machine
+   * role's policy to allow the state machine to make the API call.
+   *
+   * Use `["*"]` to grant access to all resources (discouraged).
+   *
+   * @example ["arn:aws:s3:::my_bucket"]
+   * @see https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_resource.html
+   */
+  iamResources: [string, ...string[]];
```

Comment:
I wonder if we could improve upon this by grouping all of the IAM details into a separate object?

Could the call end up looking like this?

```ts
$AWS.SDK.CloudWatch.describeAlarms({
  AlarmName: alarm.alarmName,
  iam: {
    resources: [alarm.alarmArn]
  }
});
```

By making it lower-case, we avoid any chance of collision without diverging from the AWS SDK's interface (we are strictly a super-set).

Reply:
Or (potentially too fancy), use `$`.

```ts
$AWS.SDK.CloudWatch.describeAlarms({
  AlarmName: alarm.alarmName,
  $IAM: {
    resources: [alarm.alarmArn]
  }
});
```

Reply:
Should IAM be an array by the way? Thinking of the case where an API needs access to different resources. For example, a S3 Bucket with KMS Encryption - you need access to both resources.

Also, are we losing anything by not using the grant methods? It should be possible to write your own IAM Policies and also use grant methods.

```ts
$AWS.SDK.CloudWatch.describeAlarm({
  AlarmName: alarm.alarmName,
  $IAM: [
    // option 1 - a callback
    (role) => alarm.grantDescribe(role),
    // option 2 - write a PolicyStatement
    {
       resources: [alarm.alarmArn],
       actions: [".."]
    }
  ]
});
```

Reply:
While I understand the ergonomics of merging the payloads, mixing synth and runtime will force the payload to be a object literal only for ergonomic reasons.

This wouldn't be possible in the above

```ts
$AWS.SDK.CloudWatch.describeAlarms(input.describeAlarmPayload,
{
  iam: {
    resources: [alarm.alarmArn]
  }
});
```

Reply:
Ok fair enough. I also saw the other comment about how these are L1s and grant methods belong in the L2. Fair enough.